tools/codesign_mac_executable.py

```diff
@@ -0,0 +1,232 @@
+        os.path.join(args.macos_executable, "Contents", "UE4", "Engine", "Binaries", "ThirdParty", "PhysX3", "Mac", "libApexFramework.dylib"),
+        os.path.join(args.macos_executable, "Contents", "MacOS", os.path.splitext(executable_name)[0])]
+
+    assert os.path.exists(args.entitlements_file)
+
+    for file in files:
+        cmd = ["sudo", "codesign", "-f", "-s", "-v", "--options", "runtime", "--timestamp", "--entitlements", args.entitlements_file, "--sign", f"Developer ID Application: {args.developer_id}", file]
+        print(f"[SPEAR | codesign_mac_executable.py] Executing: {' '.join(cmd)}")
+        cmd_result = subprocess.run(cmd)
+        assert cmd_result.returncode == 0
+
+    os.makedirs(args.output_dir, exist_ok=True)
+
+    # create a zip file for notarization
+    temp_zip = os.path.join(args.output_dir, f"{os.path.splitext(os.path.basename(args.macos_executable))[0]}-temp.zip")
+    cmd = ["ditto", "-c", "-k", "--rsrc", "--keepParent", args.macos_executable, temp_zip]
+    print(f"[SPEAR | codesign_mac_executable.py] Executing: {' '.join(cmd)}")
+    cmd_result = subprocess.run(cmd)
+    assert cmd_result.returncode == 0
+
+    # send the zip file for notarization 
+    cmd = ["xcrun", "altool", "--notarize-app", "--primary-bundle-id", "org.embodiedaifoundation.spear", "--username", args.apple_username, "--password", args.apple_password, "--file", temp_zip]
+    print(f"[SPEAR | codesign_mac_executable.py] Executing: {' '.join(cmd)}")
+    ps = subprocess.Popen(cmd, stdout=subprocess.PIPE)
+    output = subprocess.check_output(["grep", "RequestUUID = "], stdin=ps.stdout, text=True)
+    ps.wait()
+    ps.stdout.close()
+    request_uuid = " ".join(output[14:].split())
+    print(f"[SPEAR | codesign_mac_executable.py] Request UUID: {request_uuid}")
+
+    # check the status of notarization request
+    cmd = [
+        "xcrun", "altool", "--notarization-info", request_uuid, "--username", args.apple_username, "--password", args.apple_password]
+    print(f"[SPEAR | codesign_mac_executable.py] Executing: {' '.join(cmd)}")
+    output = "             in progress"
+    start_time = time.time()
+    elapsed_time = time.time() - start_time
+    while " ".join(output[13:].split()) == "in progress" and elapsed_time<args.wait_time_seconds:
+        ps = subprocess.Popen(cmd, stdout=subprocess.PIPE)
+        output = subprocess.check_output(["grep", "     Status:"], stdin=ps.stdout, text=True)
+        ps.wait()
+        ps.stdout.close()
+        time.sleep(20)
+        elapsed_time = time.time() - start_time
+        print(f"[SPEAR | codesign_mac_executable.py] Waiting to get more information on the notarization request...")
+    
+    assert elapsed_time <= args.wait_time_seconds, f"[SPEAR | codesign_mac_executable.py] ERROR: Exceeded provided wait time ({args.wait_time_seconds}s) for notarization. Please complete this procedure after notarization is complete."
```

Comment:
- print this error, then assert false
- since this error might happen, you should provide an option to run the end of this script, assuming the user has attempted it before and it timed out (e.g., don't try to hack the rpaths and upload the executable, just skip to this step if the user specifies a uuid)

Reply:
Done